feature (Steamworks) : SteamApiInterfaceVersionsArray

WispyMouse opened the pull request:
Starting in Steamworks v1.58a onwards, if `LibraryManagedBySDK` is defined as a Steam platform flag, then it is required to pass the version of all interface APIs in Steamworks.

This set of changes allows for definition of the Steamworks API versions inside the Steam window for EOS Configuration. You can manually insert each record, or if you're using Steamworks.NET you can press the "Update from Steamworks.NET" button to populate the list from its list of constants.

![image](https://github.com/PlayEveryWare/eos_plugin_for_unity/assets/16852552/df8ab011-7f10-4bc9-b918-bd9d48a1213c)

Facepunch users will have to manually insert these values, because those values are not exposed in a way that the plugin can programmatically pull from. It will be required for anyone who is using `LibraryManagedBySDK` moving forward to go to the EOS Configuration window once and update and save their configuration.

WispyMouse committed:
feature (Steamworks) : SteamApiInterfaceVersionsArray

As per the requirement for future steamworks version, this adds in support for pulling its Api Interface Versions, and passes it to the Native Code.

WispyMouse committed:
feature (Native) : Api Versions as array

arthur740212 commented on Assets/Plugins/Source/Editor/Utility/Steam/Steamworks_Utility.cs:
```diff
@@ -28,6 +28,7 @@
 
 #if !DISABLESTEAMWORKS
 using Steamworks;
+using System.Collections.Generic;
```
Because `GetSteamInterfacesVersions()` returns a `List`, this namespace needs to be outside of `!DISABLESTEAMWORKS` scope

WispyMouse replied:
Good catch! Changed to be outside this directive.

arthur740212 commented on Assets/Plugins/Source/Editor/Utility/Steam/Steamworks_Utility.cs:
```diff
@@ -40,6 +41,48 @@ public static string GetSteamworksVersion()
+        public static List<string> GetSteamInterfaceVersions()
+        {
+#if DISABLESTEAMWORKS
+            return string.Empty;
```
Return type mismatch if `DISABLESTEAMWORKS`
null might work here, just need to check if it affects the steam config window.

WispyMouse replied:
Oh, yup. This used to return a string and I didn't come back to change this after it became a list. Good catch!

arthur740212 reviewed (changes requested):
Great job on a new feature.
Only small picks for when Steamworks was disabled.

WispyMouse committed:
fix (Steamworks) : Using for list outside !DISABLESTEAMWORKS, return list of strings instead of string

paulhazen commented:
> Starting in Steamworks v1.58a onwards, if `LibraryManagedBySDK` is defined as a Steam platform flag, then it is required to pass the version of all interface APIs in Steamworks.
> 
> This set of changes allows for definition of the Steamworks API versions inside the Steam window for EOS Configuration. You can manually insert each record, or if you're using Steamworks.NET you can press the "Update from Steamworks.NET" button to populate the list from its list of constants.
> 
> ![image](https://private-user-images.githubusercontent.com/16852552/342456084-df8ab011-7f10-4bc9-b918-bd9d48a1213c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkzNDAyODgsIm5iZiI6MTcxOTMzOTk4OCwicGF0aCI6Ii8xNjg1MjU1Mi8zNDI0NTYwODQtZGY4YWIwMTEtN2YxMC00YmM5LWI5MTgtYmQ5ZDQ4YTEyMTNjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjI1VDE4MjYyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZhNWU5NGZlMWVhOGYwMTE4OWQ0M2I2NTUyZDE0ZTQ4NDQ0ZjJlNzM4OTMyMmRhZGE4ZmRlMzFhOGMzNGY1OGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.inhyUEgfMe4N-iYhlkVBRpCjMonIlDFT9iOem42lYXU)
> 
> Facepunch users will have to manually insert these values, because those values are not exposed in a way that the plugin can programmatically pull from. It will be required for anyone who is using `LibraryManagedBySDK` moving forward to go to the EOS Configuration window once and update and save their configuration.

Just to confirm - does a user _need_ to press the "Update values" button for this to work? If so, is there a way that we can make this happen automatically instead?

WispyMouse committed:
feature (Steamworks) : Automatically determine steamApiInterfaceVersionsArray if possible on build if empty.

WispyMouse commented:
> Just to confirm - does a user _need_ to press the "Update values" button for this to work? If so, is there a way that we can make this happen automatically instead?

I hooked up something to BuildRunner to attempt to populate the list automatically. Hopefully that'll be more convenient! It only does this population if the array is not present, or empty, in the SteamConfig file.

andrew-hirata-playeveryware commented on lib/NativeCode/DynamicLibraryLoaderHelper/NativeRender/dllmain.cpp:
```diff
@@ -1244,6 +1257,37 @@ void eos_create(EOSConfig& eosConfig)
+            steamApiInterfaceVersionsAsCharArray.push_back('\0');
+            totalBytes++;
+        }
+        steamApiInterfaceVersionsAsCharArray.push_back('\0');
```
Isn't `totalBytes` always equal to `steamApiInterfaceVersionsAsCharArray.size()`?

WispyMouse replied:
I made this go through the motions and then use `.size()`, and that did indeed get the correct size! I had been using `sizeof`, and that was only counting up until the first `\0`. `size()` is functioning correctly, so we'll use that instead of counting up in the loop like this.

andrew-hirata-playeveryware commented on lib/NativeCode/DynamicLibraryLoaderHelper/NativeRender/dllmain.cpp:
```diff
@@ -1244,6 +1257,37 @@ void eos_create(EOSConfig& eosConfig)
+        // For each element in the array (each of which is a string of an api version information)
+        // iterate across each character, and at the end of a string add a null terminator \0
+        // then add one more null terminator at the end of the array
+        std::vector<char> steamApiInterfaceVersionsAsCharArray = std::vector<char>();
```
The assignment here is not needed.

WispyMouse replied:
You're right. I had assumed that I needed to set it to something before I can start managing it, but it appears as though `std::vector<>` lets you use it right away without the assignment. Learned something new about C++!

andrew-hirata-playeveryware commented on lib/NativeCode/DynamicLibraryLoaderHelper/NativeRender/dllmain.cpp:
```diff
@@ -1244,6 +1257,37 @@ void eos_create(EOSConfig& eosConfig)
+
+        for (int apiInterfaceVersionIndex = 0; apiInterfaceVersionIndex < eos_steam_config.steamApiInterfaceVersionsArray.size(); apiInterfaceVersionIndex++)
+        {
+            std::string currentFullValue = eos_steam_config.steamApiInterfaceVersionsArray[apiInterfaceVersionIndex];
```
I believe you can get away with using a `const std::string&` here instead of a `std::string` so that the string doesn't get copied.

WispyMouse replied:
Ah, I see! This makes it point at the value in the array instead of copying it in to memory. That makes sense, I'll utilize that.

WispyMouse committed:
fix (Native) : Code review feedback on cpp changes

- Removed unneeded assignment from creating steamApiInterfaceVersionsAsCharArray
- Utilize a `const std::string&` instead of a `std::string` in order to not need to copy the string in to memory again
- Use `.size()` instead of counting up the size of the array

paulhazen committed:
Merge branch 'development' into feature/SteamApiInterfaceVersionsArray
